tests/MongoDB.Bson.Tests/Specifications/bson-corpus/BsonCorpusTestRunner.cs

```diff
@@ -0,0 +1,316 @@
+                return stream.ToArray();
+            }
+        }
+
+        private string EncodeCanonicalExtendedJson(BsonDocument document)
+        {
+#pragma warning disable 618
+            var writerSettings = new JsonWriterSettings
+            {
+                OutputMode = JsonOutputMode.CanonicalExtendedJson,
+            };
+            if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
+            {
+                writerSettings.GuidRepresentation = GuidRepresentation.Unspecified;
+            }
+#pragma warning restore 618
+            var json = document.ToJson(writerSettings);
+            return json.Replace(" ", "");
+        }
+
+        private string EncodeRelaxedExtendedJson(BsonDocument document)
+        {
+#pragma warning disable 618
+            var writerSettings = new JsonWriterSettings
+            {
+                OutputMode = JsonOutputMode.RelaxedExtendedJson,
+            };
+            if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
+            {
+                writerSettings.GuidRepresentation = GuidRepresentation.Unspecified;
+            }
+#pragma warning restore 618
+            var json = document.ToJson(writerSettings);
+            return json.Replace(" ", "");
+        }
+
+        private void RunDecodeErrorsTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "bson");
+
+            var bson = BsonUtils.ParseHexString(test["bson"].AsString);
+
+            var exception = Record.Exception(() =>
+            {
+                using (var stream = new MemoryStream(bson))
+                using (var reader = new BsonBinaryReader(stream))
+                {
+                    while (!reader.IsAtEndOfFile())
+                    {
+                        _ = BsonSerializer.Deserialize<BsonDocument>(reader);
+                    }
+                }
+            });
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunParseErrorsTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "string");
+
+            var json = test["string"].AsString;
+
+            var exception = Record.Exception(() => BsonDocument.Parse(json));
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunValidTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(
+                test,
+                "type",
+                "description",
+                "canonical_bson",
+                "canonical_extjson",
+                "relaxed_extjson",
+                "degenerate_bson",
+                "degenerate_extjson",
+                "converted_bson",
+                "converted_extjson",
+                "lossy");
+
+            byte[] cB = null;
+            if (test.Contains("canonical_bson"))
```

Comment:
Consider(and below):

            if (test.TryGetValue("canonical_bson", out var canonicalBson))
            {
                cB = BsonUtils.ParseHexString(canonicalBson.AsString);
            }
it will allow you reducing a number of string constants

Reply:
Also, consider splitting this logic into different methods like (this method a bit difficult to understand):

        if (test.TryGetValue("canonical_bson", out var canonicalBson)) // or `cB`
        {
            AsserCanonicalBson(canonicalBson);
        }

Reply:
Updated code to utilize `TryGetValue`.
Done.

I fiddled with the code and can say that splitting logic into methods lowers readability because 5 more small few-lines methods created, while they are just fine being inline, which additionally correlates well with the specification.
Not done.